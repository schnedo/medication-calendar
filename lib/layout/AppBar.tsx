import { ReactElement, useState } from "react";
import {
  AppBar as MuiAppBar,
  IconButton,
  makeStyles,
  Toolbar,
  Typography,
} from "@material-ui/core";
import { ArrowBack, Menu, SvgIconComponent } from "@material-ui/icons";
import Sidebar from "./Sidebar";
import { useRouter } from "next/router";

const useStyles = makeStyles((theme) => ({
  topLeftButtonStyle: {
    marginRight: theme.spacing(2),
  },
  titleStyle: {
    flex: "1",
  },
}));

export interface AppBarUser {
  AppBar: typeof AppBar;
}

export interface RightButtonConfig {
  icon: SvgIconComponent;
  ariaLabel: string;
  onClick?: () => void;
}

export interface AppBarProps {
  title?: string;
  withBackButton?: boolean;
  rightButton?: RightButtonConfig;
}

export default function AppBar({
  title,
  withBackButton,
  rightButton,
}: AppBarProps): ReactElement {
  const [sidebarOpen, setSidebarOpen] = useState(false);
  const router = useRouter();

  const { topLeftButtonStyle, titleStyle } = useStyles();
  return (
    <>
      <MuiAppBar position={"sticky"}>
        <Toolbar>
          <IconButton
            className={topLeftButtonStyle}
            edge={"start"}
            color={"inherit"}
            onClick={
              withBackButton
                ? () => router.back()
                : () => setSidebarOpen((old) => !old)
            }
            aria-label={withBackButton ? "go back" : "open menu"}
          >
            {withBackButton ? <ArrowBack /> : <Menu />}
          </IconButton>
          <Typography className={titleStyle} variant={"h6"}>
            {title}
          </Typography>
          <IconButton
            edge={"end"}
            color={"inherit"}
            onClick={rightButton?.onClick}
            aria-label={rightButton?.ariaLabel}
          >
            {rightButton ? <rightButton.icon /> : <></>}
          </IconButton>
        </Toolbar>
      </MuiAppBar>
      <Sidebar
        open={sidebarOpen}
        onOpen={() => setSidebarOpen(true)}
        onClose={() => setSidebarOpen(false)}
      />
    </>
  );
}
