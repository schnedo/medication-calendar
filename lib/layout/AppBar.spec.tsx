import { render } from "@testing-library/react";
import AppBar from "./AppBar";
import { mock } from "@userlike/joke";
import userEvent from "@testing-library/user-event";
import { Edit } from "@material-ui/icons";

const { default: Sidebar } = mock(import("./Sidebar"));

describe("AppBar", () => {
  beforeEach(() => {
    Sidebar.mockReturnValue(<div>mocked Sidebar</div>);
  });

  it("should render correctly", async () => {
    expect.hasAssertions();

    const { container } = render(
      <AppBar
        title={"title"}
        rightButton={{ icon: Edit, ariaLabel: "edit" }}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it("should show go back button only when flag is set", async () => {
    expect.hasAssertions();

    const { getByRole, rerender } = render(<AppBar withBackButton />);

    expect(getByRole("button", { name: "go back" })).toBeInTheDocument();
    expect(() => getByRole("button", { name: "open menu" })).toThrowError();

    rerender(<AppBar />);

    expect(getByRole("button", { name: "open menu" })).toBeInTheDocument();
    expect(() => getByRole("button", { name: "go back" })).toThrowError();
  });

  it("should open Sidebar when menu button is pressed", async () => {
    expect.hasAssertions();

    const { getByRole } = render(<AppBar />);
    expect(Sidebar).toHaveBeenCalledTimes(1);
    expect(Sidebar).toHaveBeenLastCalledWith(
      expect.objectContaining({ open: false }),
      {},
    );

    userEvent.click(getByRole("button", { name: "open menu" }));
    expect(Sidebar).toHaveBeenCalledTimes(2);
    expect(Sidebar).toHaveBeenLastCalledWith(
      expect.objectContaining({ open: true }),
      {},
    );
  });

  it("should call onRightButtonClick callback when right button is pressed", async () => {
    expect.hasAssertions();

    const onClick = jest.fn();
    const { getByRole } = render(
      <AppBar rightButton={{ icon: Edit, ariaLabel: "label", onClick }} />,
    );

    userEvent.click(getByRole("button", { name: "label" }));
    expect(onClick).toHaveBeenCalledTimes(1);
  });
});
