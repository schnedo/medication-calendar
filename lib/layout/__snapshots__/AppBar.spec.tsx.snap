// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppBar should render correctly 1`] = `
<div>
  <header
    class="MuiPaper-root MuiAppBar-root MuiAppBar-positionSticky MuiAppBar-colorPrimary MuiPaper-elevation4"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <button
        aria-label="open menu"
        class="MuiButtonBase-root MuiIconButton-root makeStyles-topLeftButtonStyle-1 MuiIconButton-colorInherit MuiIconButton-edgeStart"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <h6
        class="MuiTypography-root makeStyles-titleStyle-2 MuiTypography-h6"
      >
        title
      </h6>
      <button
        aria-label="edit"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeEnd"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </header>
  <div>
    mocked Sidebar
  </div>
</div>
`;
